### DISCUSSION TITLE
Is it possible to have lock files in each workspace pacckage?

### DISCUSSION THREAD

**Discussion by aspirisen** [pnpm/pnpm#4632] [Q&A] [answered]
Hello, i wonder if it is possible to have lock files in each workspace, but do not install the dependencies like it is separate package? So when i add dependency in workspace package pnpm updates package.json in the package, update its lock file, but pnpm will install the dependency like i would run "pnpm instal" in the root.

The use case:
I have dozens of packages and each of them have its own github repo. I want to create "super" repo where i would attach all other repos as git submodules. In the "super" repo i define workspace with paths to these submodules. The issue is that "super" repo is only for development purpose, i.e. it is not supposed to have its own lock file. At the same time all the packages do no know that they can be a part of workspace and leave independently. So when i push smth in workspace package repo CI should install its dependencies from its lock file.
Thank.

**Comment by zkochan** [comment, marked as answer]
You can try to set the [shared-workspace-lockfile](https://pnpm.io/workspaces#shared-workspace-lockfile) setting to `false`. But it will also create separate node_modules for every project. So installation will be a lot slower.

Maybe you can generate your own lockfile using this library: https://github.com/pnpm/pnpm/tree/main/packages/make-dedicated-lockfile

**Reply by aspirisen** [reply]
@zkochan thanks, If i would set [shared-workspace-lockfile](https://pnpm.io/workspaces#shared-workspace-lockfile) and it will create separate node_modules, does mean that i could have i.e. two React in bundle, or the symlink will still point to the same location?

**Reply by zkochan** [reply]
No, the symlink will not point to the same location.

**Reply by linhuiw** [reply]
I think this is the newest link for [shared-workspace-lockfile](https://pnpm.io/workspaces#shared-workspace-lockfile)

**Reply by levrik** [reply]
@zkochan Would it be possible to have only a single package in the workspace with its own lockfile? I'm having a workspace with a legacy app using React 17 and multiple newer packages using React 18. Except for the legacy application, everything gets built through a single Webpack. To avoid multiple copies of React to be included in the build, I need these to symlink to the same location. But at the same time I cannot enable `shared-workspace-lockfile` due to https://github.com/pnpm/pnpm/issues/7158 as transitive dependencies define `"@types/react": "*"` in their deps and then resolve to `^18.0.0` also for the single app which depends on `"@types/react": "^17.0.0"`. This can be solved by adding `@types/react` as peer deps to these packages through `pnpm.packageExtensions` but they are a lot! I can also solve it by overriding the resolution through `tsconfig.json` but TBH I'm afraid that there are more similar cases that result in subtle bugs and aren't as noticable as a failing build in this case.

For me the perfect solution would be to have a separate lockfile for that one app and a shared for the rest. Please note that the app using a separate lockfile would still need to be able to access other packages from the workspace which are shared between the legacy and new app so simply excluding that one package from the PNPM workspace also doesn't work.

_Edit:_ `pnpm` doesn't even respects excluded packages when executing `pnpm install` inside subdirectories. It simply installs the workspace root instead of the nested excluded package.

**Reply by kamatheuska** [reply]
@levrik try pnpm deploy to have a production build of only one package dependencies https://pnpm.io/cli/deploy

**Reply by jcalfee** [reply]
link changed again: https://pnpm.io/npmrc#shared-workspace-lockfile

**Reply by tjfroll** [reply]
@zkochan I'm running into this issue as well as I convert our monorepo to pnpm from npm workspaces. In my setup, there are directories that I do not want to include in the workspace because they are independent from workspace apps and packages. Pre-migration, the structure looks like this:

```
/apps
  /workspace-app
   - package.json
  /independent-app
   - package.json
   - package-lock.json
/packages
  /workspace-pkg
   - package.json
/devtools
  /independent-tool
   - package.json
   - package-lock.json
package.json
package-lock.json // workspace lockfile
```

For local development it's fine if they're all a part of the same workspace, nobody cares if they're installing more than they need. But in CI this is a significant shift in architecture. We have dedicated pipelines for independent apps and tools which install only their dependencies and do not utilize workspace packages. If they're forced to be a part of the workspace, then we'd need to refactor our CI to have a single shared pnpm cache available to all jobs and pipelines. This might be preferable long-term, but it's a large task. It would make migration much, much easier if we could allow isolated installations beneath a workspace.

I've tried `shared-workspace-lockfile: true` but this is an enormous performance hit - far slower than our current npm runtimes. I've also tried using `deploy` to generate lockfiles for independent directories, but the requirement to use `injectWorkspacePackages: true` drastically alters the installation behavior for workspaces.

Is there any alternative here? Otherwise it seems like we won't be able to convert our independent directories to pnpm until we refactor our CI to operate as a single workspace.

**Reply by Raj1v** [reply]
Hi @tjfroll 
Did you ever find a solution for this? We are in the exact same pickle

**Reply by rubnogueira** [reply]
@Raj1v @tjfroll did you find a solution without the performance hit?

**Reply by tjfroll** [reply]
@Raj1v @rubnogueira We were able to complete the migration with the following compromises:

1. All projects (including independent apps/tools) are included in the workspace. For CI, we switched from each project installing its own dependencies to a single dependency installation job that is shared for all projects in the workspace. This resulted in a performance loss in CI for single application deploys, but an improvement for multi-application deploys.

2. Additionally, we implemented `pnpm deploy` for our Dockerized applications. This required setting `injectWorkspacePackages: true`. We did not set this for the workspace, as this resulted in an awful local development experience (see https://github.com/orgs/pnpm/discussions/9015 and https://github.com/orgs/pnpm/discussions/3938). Instead, we set this on the fly in CI (`pnpm config set inject-workspace-packages=true`) before executing `pnpm deploy`. This also required us to address a variety of `peerDependency` violations throughout our monorepo. Addressing these violations was time-consuming but did improve the overall health and consistency of our dependency definitions across our monorepo.

So, yes and no. We were able to successfully migrate and maintain a great local development experience, but we had to make compromises on our CI performance.

**Reply by rubnogueira** [reply]
@tjfroll we end up with:

- putting in the .npmrc on the root of the monorepo with: `shared-workspace-lockfile=false`
- in every project, put `recursive-install=false` on every project.
- having everything set in the `pnpm-workspace.yaml`

With that, every time we did the pnpm install, no matter the location, we had independent locks and the pnpm install could be managed by the monorepo side as well.

**Comment by dcaillibaud** [comment]
I have a similar problem with a workspace containing an app and a lot of packages it use.

I would like a pnpm-lock.yaml for the app, to allow deploying it with exactly the same versions of its dependencies, but I don't neet it for all others packages.

Adding a `.npmrc` file à the root of the workspace with `shared-workspace-lockfile=false` do the job for the app, but also for all packages, and it's really slower.

[make-dedicated-lockfile](https://github.com/pnpm/pnpm/tree/main/packages/make-dedicated-lockfile#pnpmmake-dedicated-lockfile) works too, but it's deprecated, so I guess it will disapear in next pnpm version.

And `pnpm deploy` isn't an option for me (because of my deployment workflow, build is done on the targeted server with the git repo as source).

**Comment by Goldziher** [comment]
Same issue. `shared-workspace-lockfile=false` works, but is indeed slower. 

I don't know how difficult it is to improve the speed of.

**Comment by jcalfee** [comment]
I require pnpm in cases where components depend on those links so core packages are not double loaded but this can't scale well without the really locking things down (package lock).  For for getting CI/CD and reliable builds, I'm realizing I have to do this outside of pnpm (at least for now).  So, I'm leaning towards our own https://bit.cloud/ server and npm registry server.  It has many features I really need anyways.  Although I have not gone very far down this path, I'm pretty sure this is the way to go.

**Comment by cefn** [comment]
Rather than changing lockfile format, I think an effective workaround for us would be to detect workspace packages that have arrived at any different lockfile resolutions since e.g. `origin/main`.

Then we can selectively re-test any ___affected___ packages while still having a global lockfile.

Basically if you make a lockfile edit that broke other packages this would be detected (since they all re-run their builds). If you made a too-broad-change (e.g. in a monorepo deleting the lockfile), you could back this out by checking out `origin/main` lockfile again and making a cleaner change.

Is there a pnpm command to query changed resolutions, that respects the commit-ish logic of pnpm to compare `main`?
